README
README file for Graph::Easy::Marpa.

See also: CHANGES and Changelog.ini.

Warning: WinZip 8.1 and 9.0 both contain an 'accidental' bug which stops
them recognizing POSIX-style directory structures in valid tar files.
You are better off using a reliable tool such as InfoZip:
ftp://ftp.info-zip.org/pub/infozip/

1 Installing from a Unix-like distro
------------------------------------
shell>gunzip Graph-Easy-Marpa-1.00.tgz
shell>tar mxvf Graph-Easy-Marpa-1.00.tar

On Unix-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>./Build
shell>./Build test
shell>./Build install

On MS Windows-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>perl Build
shell>perl Build test
shell>perl Build install

Alternately, without Module::Build, you do this:

Note: 'make' on MS Windows-like systems may be called 'nmake' or 'dmake'.

shell>perl Makefile.PL
shell>make
shell>make test
shell>su              (for Unix-like systems)
shell>make install
shell>exit            (for Unix-like systems)

On all systems:

Run Marpa.pm through your favourite pod2html translator.

2 Installing from an ActiveState distro
---------------------------------------
shell>unzip Graph-Easy-Marpa-1.00.zip
shell>ppm install --location=. Graph-Easy-Marpa
shell>del Graph-Easy-Marpa-1.00.ppd
shell>del PPM-Graph-Easy-Marpa-1.00.tar.gz


CHANGES
Revision history for Perl extension Graph::Easy::Marpa.

1.12  Thu Dec 20 15:27:00 2012
	- Change all uses of shape 'square' to 'rect', since older versions of Graphviz didn't have 'square'.
		This affects data/class.group.04.raw, data/class.node.09.raw, and POD in Marpa.pm and Lexer.pm
		The corresponding data/*.cooked files change as a result, when they are re-generated.
	- Change all functions (not methods) in Parser.pm from my(undef...) to my($stash...).
		This feature of Marpa is not used, but might be one day. The $stash is a hashref of per-parse info.
	- Change sub log():
		From: $self -> logger -> $level($s)        if ($self -> logger)
		To:   $self -> logger -> log($level => $s) if ($self -> logger)
		The reason for this is that if your logger is derived from another class or role, that class or
		role had to provide not just the log() method, but all of debug(), info(), notice(), warning(),
		warn(), error(), err(), critical(), crit(), alert(), emergency() and emerg(), just in case any
		of them are needed.
		I should not have used that design, because it forces other code to provide too much complexity.
		This affects Graph::Easy::Marpa, Graph::Easy::Marpa::Lexer, Graph::Easy::Marpa::Parser and
		Graph::Easy::Marpa::Renderer::GraphViz2.

1.11  Thu Nov  8 12:38:00 2012
	- No user-visible code changes.
	- For pre-reqs such as strict, warnings, etc, which ship with Perl, set the version # to 0.
		Reported as RT#80663 by Father Chrysostomos for Tree::DAG_Node.
	- Remove color cycle code from scripts/stt.html, since it was never used.
	- Rename HTML templates (to match) from rainbow.table.tx to basic.table.tx and rainbow.row.tx to basic.column.tx.
	- Rename rainbow.page.tx to stt.tx.
	- Move html/graph.easy.index.tx to htdocs/assets/templates/graph/easy/marpa.
	- Add Graph::Easy::Marpa::Config to manage location of templates.
	- Add config/.htgraph.easy.marpa.conf.
	- Add scripts/copy.config.pl.
	- Move code from scripts/stt.pl into Graph::Easy::Marpa::Utils.
	- Ship html/stt.html.
	- Add environment report to HTML templates.
	- Link graph demo template and stt template to each other.
	- Update pre-reqs.
	- Remove demo files which contained syntax errors: data/graph.10.raw, data/graph.12.raw and data/node.11.raw.

1.10  Tue Oct 16 09:49:00 2012
	- Fix a bug in the lexer where it used to combine multiple lines in the input file by, basically, using
		join('', <IN>), which meant the last char of a line was juxtaposed with the first char of the next
		line. If the second line started with spaces, that was ok, but if it started with, say, the name of
		a node, then that name was concatenated with what could have been a node name on the end of a previous
		line. This meant 2 separate node names sometimes became one.

1.09  Mon Jun 25 14:10:00 2012
	- Handle HTML-style labels, even if they contain ';' chars, e.g. by using HTML entities such as &lt;.
	- Add matching test files data/node.(17, 18, 19).(raw, cooked). See also the demo page.
	- Change the abstract slightly to mention we process Graphviz files in the Graph::Easy format.
	- Expand the FAQ re HTML-style labels.
	- Encode HTML entities when including the demos' source code in the demo page, because now demos can
		contain HTML-like labels.

1.08  Sun Dec 25 10:33:00 2011
	  - Change <img... to <object...> in the demo creation code, to keep poor old FireFox happy.
	  - Change various things in html/graph.easy.index.tx to we validate as XHTML 1.0 Strict.
	  - Switch from Marpa::XS to Marpa::R2.

1.07  Wed Dec 14 13:29:00 2011
	  - Fix declaration of null_value for some rules, to stop heaps of warnings during testing and/or running gem.pl.
	  	The messages you get look like: Zero length sequence for symbol without null value: "group_name",
		if you don't have symbol descriptors declared for symbols whose value can be empty.
		Search Graph::Easy::Marpa::Parser for 'symbols' to see what I had to do because group and node names can be empty.
	  - Update Build.PL and Makefile.PL to only require Marpa::XS V 0.026000 (not 0.108000), which is the latest as of today.
	  - Add a discussion of Marpa::PP to the FAQ, since I've tested it too.

1.06  Mon Aug  8 11:08:00 2011
	  - Improve home-grown validate.build.make.pl to include not just *.pm files, but also *.pl and *.t files, so as to find more dependencies.
	  - Add Capture::Tiny, File::Remove, File::Temp, Getopt::Long, List::Cycle, Pod::Usage and Text::Xlate to Build.PL and Makefile.PL.
	  - And yes I know, List::Cycle is not really needed (yet) :-).
	  - Patch the STT so that attribute definitions for classes can have leading spaces, as edge and node attribute definitions already could.
	  	Eg: 'graph {rankdir: TB }' can now be 'graph { rankdir: TB}'. Note the new space in front of 'rankdir'.

1.05  Tue Aug  2 15:32:00 2011
	  - Add File::Slurp to Build.PL and Makefile.PL because its readdir() method is used in t/cooked.t and t/raw.t.
	  - Add http://metacpan.org/ link to html/graph.easy.index.tx, so it appears in html/index.html.
	  - Clean up confusion between demo output being png or svg. The demo files uploaded to my web site used to be png,
	  	because one of the outputs (graph.15.*) was empty (for some reason) when svg is used.
		But since svg files are smaller than png files, and everything works, svg is used again.
	  - Add DOCTYPE and Content-Type to html/graph.easy.index.tx, so that html/index.html validates as XHTML 1.0 Transitional.

1.04  Fri Jul 29 14:12:00 2011
	  - Make scripts/gem.pl pass rankdir thru to parser.
	  - Make scripts/generate.demo.sh pass image type thru to scripts/generate.demo.pl.
	  - Fix scripts/graph.easy.index.tx img src field patch which got lost with a git reset.

1.03  Sat Jul 23 16:36:00 2011
	  - Add Sort::Naturally to Build.PL and Makefile.PL, since it's used in t/*.t.

1.02  Thu Jul 21 13:37:00 2011
	  - Change colors so they're compatible with both the X11 and SVG color schemes.

Makefile.PL
use ExtUtils::MakeMaker;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

WriteMakefile
(
	($] ge '5.005') ?
	(
		AUTHOR   => 'Ron Savage (ron@savage.net.au)',
		ABSTRACT => 'A Marpa-based parser for Graph::Easy-style Graphviz files',
	) : (),
	clean =>
	{
		FILES => 'blib/* Makefile MANIFEST Graph-Easy-Marpa-*'
	},
	dist =>
	{
		COMPRESS => 'gzip',
		SUFFIX   => 'gz'
	},
	DISTNAME  => 'Graph-Easy-Marpa',
	NAME      => 'Graph::Easy::Marpa',
	PL_FILES  => {},
	PREREQ_PM =>
	{
		Capture::Tiny         => 0.11,
		Config                => undef,
		Config::Tiny          => 2.14,
		Data::Section::Simple => 0.02,
		Date::Simple          => 3.03,
		File::Basename        => 2.78,
		File::Copy            => 2.21,
		File::HomeDir         => 0.99,
		File::Slurp           => 9999.19,
		File::Spec            => 3.33,
		Getopt::Long          => 2.38,
		GraphViz2             => 1.09,
		Hash::FieldHash       => 0.10,
		HTML::Entities::Interpolate => 1.04,
		IO::File              => 1.14,
		List::Compare         => 0.37,
		Log::Handler          => 0.71,
		Marpa::R2             => 2.022000,
		Module::Load          => 0.18,
		#OpenOffice::OODoc    => 2.125, # Optionally used in Lexer.pm.
		Path::Class           => 0.24,
		Perl6::Slurp          => 0.03,
		Pod::Usage            => 1.36,
		Set::Array            => 0.23,
		Set::FA::Element      => 1.08,
		Sort::Naturally       => 1.02,
		strict                => 0,
		Test::More            => 0.94,
		Test::Pod             => 1.44,
		Text::CSV_XS          => 0.80,
		Text::Xslate          => 1.4001,
		Try::Tiny             => 0.09,
		warnings              => 0,
	},
	VERSION_FROM => 'lib/Graph/Easy/Marpa.pm',
);
